cmd/tk/export.go

```diff
@@ -1,137 +1,87 @@
 package main
 
 import (
-	"bytes"
 	"fmt"
-	"io"
-	"io/ioutil"
-	"log"
-	"os"
-	"path/filepath"
-	"strings"
 
-	"text/template"
-
-	"github.com/Masterminds/sprig/v3"
 	"github.com/go-clix/cli"
+	"k8s.io/apimachinery/pkg/labels"
 
 	"github.com/grafana/tanka/pkg/tanka"
 )
 
-// BelRune is a string of the Ascii character BEL which made computers ring in ancient times
-// We use it as "magic" char for the subfolder creation as it is a non printable character and thereby will never be
-// in a valid filepath by accident. Only when we include it.
-const BelRune = string(rune(7))
-
 func exportCmd() *cli.Command {
 	args := workflowArgs
-	args.Validator = cli.ValidateExact(2)
+	args.Validator = cli.ValidateFunc(func(args []string) error {
+		if len(args) < 2 {
+			return fmt.Errorf("expects at least 2 args, received %v", len(args))
+		}
+		return nil
+	})
 
 	cmd := &cli.Command{
-		Use:   "export <environment> <outputDir>",
-		Short: "write each resources as a YAML file",
+		Use:   "export <outputDir> <path> [<path>...]",
+		Short: "export environments found in path(s)",
 		Args:  args,
 	}
 
-	format := cmd.Flags().String("format", "{{.apiVersion}}.{{.kind}}-{{.metadata.name}}", "https://tanka.dev/exporting#filenames")
-	extension := cmd.Flags().String("extension", "yaml", "File extension")
-	merge := cmd.Flags().Bool("merge", false, "Allow merging with existing directory")
+	defaultOpts := tanka.DefaultExportEnvOpts()
+
+	format := cmd.Flags().String("format", defaultOpts.Format, "https://tanka.dev/exporting#filenames")
+	dirFormat := cmd.Flags().String("dirformat", defaultOpts.DirFormat, "based on tanka.dev/Environment object")
+
+	extension := cmd.Flags().String("extension", defaultOpts.Extension, "File extension")
+	merge := cmd.Flags().Bool("merge", defaultOpts.Merge, "Allow merging with existing directory")
+	recursive := cmd.Flags().BoolP("recursive", "r", false, "Look recursively for Tanka environments")
+	labelSelector := cmd.Flags().StringP("selector", "l", "", "Label selector. Uses the same syntax as kubectl does")
```

Comment:
In L27 you initialize a variable of type `ExportEnvOpts`, which you use for gathering default values.
Also in L73 and following you create another field of same type to pass to the actual function call.

This is an opportunity to use `cmd.Flags().StringVar`, etc. to bind the flags to the `defaultOpts` variable, so you only have one

Reply:
I took the opportunity. :wink: